src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp

```diff
@@ -0,0 +1,432 @@
+// Copyright 2019 ADLINK Technology
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef DDSCXXDATATOPIC_HPP_
+#define DDSCXXDATATOPIC_HPP_
+
+#include <memory>
+#include <string>
+#include <cstring>
+#include <vector>
+
+#include "dds/ddsrt/endian.h"
+#include "dds/ddsi/q_radmin.h"
+#include "dds/ddsi/ddsi_serdata.h"
+#include "dds/ddsi/ddsi_sertopic.h"
+
+enum class endianness
+{
+  little_endian = DDSRT_LITTLE_ENDIAN,
+  big_endian = DDSRT_BIG_ENDIAN
+};
+
+constexpr endianness native_endianness() { return endianness(DDSRT_ENDIAN); }
+
+static inline void* calc_offset(void* ptr, ptrdiff_t n)
+{
+  return static_cast<void*>(static_cast<unsigned char*>(ptr) + n);
+}
+
+static inline const void* calc_offset(const void* ptr, ptrdiff_t n)
+{
+  return static_cast<const void*>(static_cast<const unsigned char*>(ptr) + n);
+}
+
+template <typename T>
+class ddscxx_sertopic : public ddsi_sertopic {
+public:
+  static const struct ddsi_sertopic_ops ddscxx_sertopic_ops;
+  ddscxx_sertopic(const char* topic_name, const char* type_name);
+};
+
+template <typename T>
+class ddscxx_serdata : public ddsi_serdata {
+  size_t m_size{ 0 };
+  std::unique_ptr<unsigned char[]> m_data{ nullptr };
+
+public:
+  static const struct ddsi_serdata_ops ddscxx_serdata_ops;
+  ddscxx_serdata(const ddsi_sertopic* topic, ddsi_serdata_kind kind);
+
+  void resize(size_t requested_size);
+  size_t size() const { return m_size; }
+  void* data() const { return m_data.get(); }
+};
+
+template <typename T>
+bool serdata_eqkey(const struct ddsi_serdata* a, const struct ddsi_serdata* b)
+{
+  (void)a;
+  (void)b;
+  return true;
```

Comment:
Don't know how much of this is actually required, but it kind of seems that keys aren't compared and just always match? Am I correct in my assumption?

Reply:
At the moment this code was written, the key generation was not yet in the generated code, it now is.
I will replace these with the correct key calculation functions.